tokio-macros/src/entry.rs

```diff
@@ -369,10 +370,15 @@ fn parse_knobs(mut input: ItemFn, is_test: bool, config: FinalConfig) -> TokenSt
     };
 
     let body_ident = quote! { body };
+    let return_or_nothing = match &input.sig.output {
+        syn::ReturnType::Type(_, ty) if matches!(**ty, syn::Type::Never(_)) => quote! {},
+        _ => quote! { return },
+    };
+
     let last_block = quote_spanned! {last_stmt_end_span=>
         #[allow(clippy::expect_used, clippy::diverging_sub_expression)]
         {
-            return #rt
+            #return_or_nothing #rt
```

Comment:
If we always emit `return`, the following code will result to compiler warning(unreachable statement for `return`)

macros_test.rs: 36
```rust
// https://github.com/tokio-rs/tokio/issues/4175
#[tokio::main]
pub async fn issue_4175_main_1() -> ! {
    panic!();
}```

So with this PR, the rustc properly detects that the `return` clause is meaningless for `!` type returns.